pymc/gp/cov.py

```diff
@@ -548,7 +568,48 @@ def full(self, X, Xs=None):
         return pt.exp(-0.5 * r)
 
 
-class ExpQuad(Stationary):
+class GeneralizedPeriodic(Stationary):
+    r"""
+    The Generalized Periodic kernel.
+
+    Can be used to wrap any isotropic stationary kernel to transform it
+    into a periodic version. The derivation can be achieved by mapping the
+    original inputs through the transformation u = (cos(x), sin(x)).
+    """
+
+    def __init__(self, base_kernel: IsotropicStationary, period):
+        self.base_kernel = base_kernel
+        self.period = period
+
+    @property
+    def input_dim(self):
+        return self.base_kernel.input_dim
+
+    @property
+    def active_dims(self):
+        return self.base_kernel.active_dims
+
+    @property
+    def ls(self):
+        return self.base_kernel.ls
+
+    def full(self, X, Xs=None):
+        X, Xs = self._slice(X, Xs)
+        if Xs is None:
+            Xs = X
+        f1 = X.dimshuffle(0, "x", 1)
+        f2 = Xs.dimshuffle("x", 0, 1)
+        r = np.pi * (f1 - f2) / self.period
+        if hasattr(self.base_kernel, "full_r"):
+            r = pt.sum(pt.abs(pt.sin(r) / self.ls), 2)
```

Comment:
Should we keep the multiplying 4 or not? I see that GPflow chose to leave it out and have it absorbed by the `ls`.

My ruff derivation of the argument that goes into `r` is

$$
y = 2\pi x / period
$$

$$
y' = 2\pi x' / period
$$

$$
r^2 = (\cos(y) - \cos(y'))^2 + (\sin(y) - \sin(y'))^2
$$

$$
r^2 = 2 - 2\cos(y - y')
$$

$$
r^2 = 2 \left[sin\left(\frac{y - y'}{2}\right)^2 + cos\left(\frac{y - y'}{2}\right)^2 - \cos(y - y') \right]
$$

$$
r^2 = 4 sin\left(\frac{y - y'}{2}\right)^2
$$

$$
r^2 = 4 sin\left(\frac{\pi(x - x')}{period}\right)^2
$$

From looking at `Periodic` it seems as though we had dropped it long ago, so it might be irrelevant

Reply:
I'd prob just go with what is consistent with current implementation.